quil-py/src/program/mod.rs

```diff
@@ -218,8 +225,75 @@ impl PyProgram {
             .to_owned())
     }
 
-    pub fn __str__(&self) -> String {
-        self.as_inner().to_string()
+    pub fn resolve_placeholders(&mut self) {
+        self.as_inner_mut().resolve_placeholders();
+    }
+
+    // TODO: Pre-resolve placeholders into a map so a non-panic error can be raised?
+    #[args("*", label_resolver = "None", qubit_resolver = "None")]
+    pub fn resolve_placeholders_with_custom_resolvers(
+        &mut self,
+        label_resolver: Option<Py<PyFunction>>,
+        qubit_resolver: Option<Py<PyFunction>>,
+    ) {
+        #[allow(clippy::type_complexity)]
+        let rs_qubit_resolver: Box<dyn Fn(&QubitPlaceholder) -> Option<u64>> =
+            if let Some(resolver) = qubit_resolver {
+                Box::new(move |placeholder: &QubitPlaceholder| -> Option<u64> {
+                    Python::with_gil(|py| {
+                        let resolved_qubit =
+                            resolver.call1(py, (placeholder.to_python(py).unwrap(),));
+                        assert!(
+                            resolved_qubit.is_ok(),
+                            "qubit_resolver returned an error: {resolved_qubit:?}"
+                        );
+
+                        let resolved_qubit: PyResult<Option<u64>> = resolved_qubit
+                            .expect("asserted that resolved_qubit is ok")
```

Comment:
I suppose you can just `unwrap()` here since, if there is a problem, it will have been caught above (i.e. this `expect` will never fail).